sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -414,4 +352,269 @@ trait CheckAnalysis extends PredicateHelper {
 
     plan.foreach(_.setAnalyzed())
   }
+
```

Comment:
note to reviewers: This function basically refactors the validation logic for subquery expressions from checkAnalysis. This is the entry point function to do all the validation for subquery is is called from checkAnalysis().